torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -165,8 +165,38 @@ void RRefContext::delUser(
 
     fm->addCallback([](const Message& /* unused */,
                        const c10::optional<utils::FutureError>& futErr) {
-      RRefContext::handleException(futErr);
+      handleException(futErr);
     });
+
+    confirmedUsers_.erase(forkId);
+  }
+}
+
+void RRefContext::delAllUsers(std::chrono::milliseconds timeoutMillis) {
+  // First, wait for all pending UserRRefs to be confirmed,
+  // one kind is pendingUsers_, which are shared from Owner,
+  // the other kind pendingChildren_, which are shared from another User.
+  std::unique_lock<std::mutex> lock(mutex_);
+  bool noPendingAndOwner =
+      pendingReducedCV_.wait_for(lock, timeoutMillis, [this]() {
+        return pendingUsers_.size() == 0 && pendingChildren_.size() == 0 &&
+            owners_.size() == 0;
```

Comment:
don't we need to wait for owners after deleting all users? I mean sth like:

1. wait for pendingUsers_ and pendingChildren_
2. delete all users 
3. wait for all owners_?

Otherwise, A could be holding a user rref of B, while B could holding a user rref to A? and neither of them would proceed?

Reply:
I tried making that change, but could still hit the following error. Hmm...

```
Traceback (most recent call last):
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/data/users/shenli/pytorch/torch/testing/_internal/common_distributed.py", line 243, in _run
    getattr(self, test_name)()
  File "/data/users/shenli/pytorch/torch/testing/_internal/dist_utils.py", line 71, in new_test_method
    rpc.shutdown(graceful=clean_shutdown)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 77, in wrapper
    return func(*args, **kwargs)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 240, in shutdown
    _get_current_rpc_agent().join()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:572] Connection closed by peer [2401:db00:21:6058:face:0:3:0]:5529
ERROR:torch.distributed.rpc.api:worker2 failed to respond to 'Shutdown Proceed.' request in 0:00:05
Process process 0:
Traceback (most recent call last):
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/data/users/shenli/pytorch/torch/testing/_internal/common_distributed.py", line 243, in _run
    getattr(self, test_name)()
  File "/data/users/shenli/pytorch/torch/testing/_internal/dist_utils.py", line 71, in new_test_method
    rpc.shutdown(graceful=clean_shutdown)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 77, in wrapper
    return func(*args, **kwargs)
  File "/data/users/shenli/pytorch/torch/distributed/rpc/api.py", line 240, in shutdown
    _get_current_rpc_agent().join()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:572] Connection closed by peer [2401:db00:21:6058:face:0:3:0]:38794
```

Reply:
The problem is that I didn't wait for response of delete user rref.
So user could run to `_agent.join()` right after sending delete user rref request, while owner is also waiting on `_agent.join()`. At this time, owner can destroy rref context and report owner rref leaks. After this, the server thread picks up the request to delete user rref and realized no such rref id exists and the owner node crashes. Then, User node detects GLOO connection cloed by peer.

Reply:
"wait for all owners_" should be interchangeable with "agent.join()", if client waits for response.